crates/hotshot/src/traits/networking/push_cdn_network/definition/message_hook.rs

```diff
@@ -0,0 +1,577 @@
+    /// from reconnecting and using a brand new GCR instance (empty rate limit)
+    dropped_gcr_cache: Arc<Mutex<LruCache<u64, (Gcr, Gcr)>>>,
+
+    /// The global average number of bytes per period for direct messages
+    global_average_direct_bps: Arc<Average>,
+
+    /// The global average number of bytes per period for broadcast messages
+    global_average_broadcast_bps: Arc<Average>,
+
+    /// The local GCR instance for direct messages
+    local_gcr_direct: Gcr,
+
+    /// The local GCR instance for broadcast messages
+    local_gcr_broadcast: Gcr,
+
+    /// The local, running number of consumed bytes for direct messages
+    local_sample_direct_bps: Sample,
+
+    /// The local, running number of consumed bytes for broadcast messages
+    local_sample_broadcast_bps: Sample,
+}
+
+impl Default for HotShotMessageHook {
+    /// # Panics
+    /// If 100 < 0
+    fn default() -> Self {
+        Self::new(Duration::from_secs(60), u32::MAX)
+    }
+}
+
+impl HotShotMessageHook {
+    /// Create a new `HotShotMessageHook`
+    ///
+    /// # Panics
+    /// If 100 < 0
+    #[must_use]
+    pub fn new(period: Duration, starting_rate: u32) -> Self {
+        Self {
+            message_hash_cache: LruCache::new(NonZeroUsize::new(100).unwrap()),
+            num_hooks: Arc::new(()),
+            identifier: rand::random(),
+            dropped_gcr_cache: Arc::new(Mutex::new(LruCache::new(
+                NonZeroUsize::new(1000).unwrap(),
+            ))),
+            commit_interval: period,
+            local_gcr_direct: Gcr::new(starting_rate, period, Some(u32::MAX)).unwrap(),
+            local_gcr_broadcast: Gcr::new(starting_rate, period, Some(u32::MAX)).unwrap(),
+            local_sample_direct_bps: Sample::new(),
+            local_sample_broadcast_bps: Sample::new(),
+            global_average_direct_bps: Arc::new(Average::new()),
+            global_average_broadcast_bps: Arc::new(Average::new()),
+        }
+    }
+
+    /// Process a message against the moving average
+    /// Returns whether or not the message should be skipped.
+    fn process_against_average(
+        &mut self,
+        message_len: usize,
+        message_type: MessageType,
+    ) -> HookResult {
+        // Get the correct sample, `GCR`, and global average based on the message type
+        let (local_sample, gcr, global_average) = match message_type {
+            MessageType::Broadcast => (
+                &mut self.local_sample_broadcast_bps,
+                &mut self.local_gcr_broadcast,
+                &self.global_average_broadcast_bps,
+            ),
+            MessageType::Direct => (
+                &mut self.local_sample_direct_bps,
+                &mut self.local_gcr_direct,
+                &self.global_average_direct_bps,
+            ),
+        };
+
+        // Check against the `GCR` instance, skipping if we're rate limited
+        // If we hit a parameter error, we'll process the message anyway
```

Comment:
A parameter error is internal to `Gcr`, it can be hit if, say, you try to update the average to zero. This makes it so we don't accidentally block all messages everywhere if there was an issue with that